File: statefun-examples/statefun-ridesharing-example/statefun-ridesharing-example-simulator/src/main/java/org/apache/flink/statefun/examples/ridesharing/simulator/simulation/Driver.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.flink.statefun.examples.ridesharing.simulator.simulation;

import com.google.common.base.Preconditions;
import java.util.concurrent.ThreadLocalRandom;
import javax.annotation.Nullable;
import org.apache.flink.statefun.examples.ridesharing.generated.InboundDriverMessage;
import org.apache.flink.statefun.examples.ridesharing.generated.OutboundDriverMessage;
import org.apache.flink.statefun.examples.ridesharing.simulator.model.WebsocketDriverEvent;
import org.apache.flink.statefun.examples.ridesharing.simulator.simulation.engine.LifecycleMessages;
import org.apache.flink.statefun.examples.ridesharing.simulator.simulation.engine.Simulatee;

public class Driver implements Simulatee {
  private final String driverId;
  private final DriverMessaging messaging;
  private final StateMachine<WebsocketDriverEvent.DriverStatus> stateMachine;
  private final int gridDimension;

  private int currentLocation;

  @Nullable private WebsocketDriverEvent.RideInformation rideInformation;

  Driver(String driverId, DriverMessaging messaging, int gridDimension, int startLocation) {
    this.driverId = driverId;
    this.messaging = messaging;
    this.gridDimension = gridDimension;
    this.currentLocation = startLocation;

    stateMachine = new StateMachine<>(WebsocketDriverEvent.DriverStatus.IDLE);

    // we don't have anything to do on initialization
    stateMachine.withState(
        WebsocketDriverEvent.DriverStatus.IDLE,
        LifecycleMessages.Initialization.class,
        unused -> WebsocketDriverEvent.DriverStatus.IDLE);

    // idle heartbeat
    stateMachine.withState(
        WebsocketDriverEvent.DriverStatus.IDLE, LifecycleMessages.TimeTick.class, this::heartbeat);

    // pickup request
    stateMachine.withState(
        WebsocketDriverEvent.DriverStatus.IDLE,
        OutboundDriverMessage.class,
        OutboundDriverMessage::hasPickupPassenger,
        this::pickupPassenger);

    stateMachine.withState(
        WebsocketDriverEvent.DriverStatus.PICKUP,
        LifecycleMessages.TimeTick.class,
        this::preformPickup);
    stateMachine.withState(
        WebsocketDriverEvent.DriverStatus.ENROUTE,
        LifecycleMessages.TimeTick.class,
        this::preformRoute);
  }

  /** send periodic heart beats when idle */
  private WebsocketDriverEvent.DriverStatus heartbeat(
      @SuppressWarnings("unused") LifecycleMessages.TimeTick ignored) {

    // notify application
    messaging.outgoingDriverEvent(
        InboundDriverMessage.newBuilder()
            .setDriverId(driverId)
            .setLocationUpdate(
                InboundDriverMessage.LocationUpdate.newBuilder().setCurrentGeoCell(currentLocation))
            .build());

    // notify the websocket
    messaging.broadcastDriverSimulationEvent(
        WebsocketDriverEvent.builder()
            .currentLocation(currentLocation)
            .driverId(driverId)
            .driverStatus(WebsocketDriverEvent.DriverStatus.IDLE)
            .ride(null)
            .build());

    return WebsocketDriverEvent.DriverStatus.IDLE;
  }

  /** receive a pickup command and start riding to the passenger */
  private WebsocketDriverEvent.DriverStatus pickupPassenger(OutboundDriverMessage message) {
    OutboundDriverMessage.PickupPassenger pickup = message.getPickupPassenger();

    // capture ride info from the pickup message
    this.rideInformation =
        WebsocketDriverEvent.RideInformation.builder()
            .passengerId(pickup.getRideId()) // TODO: fix this at the application side.
            .pickupLocation(pickup.getStartGeoLocation())
            .dropoffLocation(pickup.getEndGeoLocation())
            .build();

    // notify the websocket
    messaging.broadcastDriverSimulationEvent(
        WebsocketDriverEvent.builder()
            .ride(rideInformation)
            .driverStatus(WebsocketDriverEvent.DriverStatus.PICKUP)
            .driverId(driverId)
            .currentLocation(currentLocation)
            .build());

    return WebsocketDriverEvent.DriverStatus.PICKUP;
  }

  private WebsocketDriverEvent.DriverStatus preformPickup(
      @SuppressWarnings("unused") LifecycleMessages.TimeTick ignored) {
    Preconditions.checkState(rideInformation != null, "should have ride information.");
    if (currentLocation == rideInformation.getPickupLocation()) {
      // we have reached to the passenger, lets pick him up!

      messaging.broadcastDriverSimulationEvent(
          WebsocketDriverEvent.builder()
              .currentLocation(currentLocation)
              .driverId(driverId)
              .driverStatus(WebsocketDriverEvent.DriverStatus.ENROUTE)
              .ride(rideInformation)
              .build());

      messaging.outgoingDriverEvent(
          InboundDriverMessage.newBuilder()
              .setDriverId(driverId)
              .setRideStarted(InboundDriverMessage.RideStarted.getDefaultInstance())
              .build());

      return WebsocketDriverEvent.DriverStatus.ENROUTE;
    }

    // we need to advance toward the passenger
    int selfX = currentLocation / gridDimension;
    int selfY = currentLocation % gridDimension;
    final int passX = rideInformation.getPickupLocation() / gridDimension;
    final int passY = rideInformation.getPickupLocation() % gridDimension;

    if (ThreadLocalRandom.current().nextBoolean()) {
      //
      // advance in X
      //
      if (selfX > passX) {
        selfX--;
      } else if (selfX < passX) {
        selfX++;
      }
    } else {
      //
      // advance in Y
      //
      if (selfY > passY) {
        selfY--;
      } else if (selfY < passY) {
        selfY++;
      }
    }

    currentLocation = selfX * gridDimension + selfY;

    // send a heartbeat with our new location
    // notify application of our new location
    messaging.outgoingDriverEvent(
        InboundDriverMessage.newBuilder()
            .setDriverId(driverId)
            .setLocationUpdate(
                InboundDriverMessage.LocationUpdate.newBuilder().setCurrentGeoCell(currentLocation))
            .build());

    // notify the websocket
    messaging.broadcastDriverSimulationEvent(
        WebsocketDriverEvent.builder()
            .currentLocation(currentLocation)
            .driverId(driverId)
            .driverStatus(WebsocketDriverEvent.DriverStatus.PICKUP)
            .ride(rideInformation)
            .build());

    return WebsocketDriverEvent.DriverStatus.PICKUP;
  }

  private WebsocketDriverEvent.DriverStatus preformRoute(
      @SuppressWarnings("unused") LifecycleMessages.TimeTick ignored) {
    Preconditions.checkState(rideInformation != null, "should have ride information.");
    if (currentLocation == rideInformation.getDropoffLocation()) {
      // done!

      // notify websocket
      messaging.broadcastDriverSimulationEvent(
          WebsocketDriverEvent.builder()
              .currentLocation(currentLocation)
              .driverId(driverId)
              .driverStatus(
                  WebsocketDriverEvent.DriverStatus
                      .RIDE_COMPLETED) // TODO: should we send ride done?
              .ride(rideInformation)
              .build());

      // notify application
      messaging.outgoingDriverEvent(
          InboundDriverMessage.newBuilder()
              .setDriverId(driverId)
              .setRideEnded(
                  InboundDriverMessage.RideEnded.newBuilder()
                      .setRideId(rideInformation.getPassengerId()))
              .build());

      rideInformation = null;

      // we switch back to idle
      return WebsocketDriverEvent.DriverStatus.IDLE;
    }

    // we need to advance toward the passenger's dropoff location
    int selfX = currentLocation / gridDimension;
    int selfY = currentLocation % gridDimension;
    final int passX = rideInformation.getDropoffLocation() / gridDimension;
    final int passY = rideInformation.getDropoffLocation() % gridDimension;

    if (ThreadLocalRandom.current().nextBoolean()) {
      //
      // advance in X
      //
      if (selfX > passX) {
        selfX--;
      } else if (selfX < passX) {
        selfX++;
      }
    } else {
      //
      // advance in Y
      //
      if (selfY > passY) {
        selfY--;
      } else if (selfY < passY) {
        selfY++;
      }
    }

    currentLocation = selfX * gridDimension + selfY;

    // send a heartbeat with our new location
    // notify application of our new location
    messaging.outgoingDriverEvent(
        InboundDriverMessage.newBuilder()
            .setDriverId(driverId)
            .setLocationUpdate(
                InboundDriverMessage.LocationUpdate.newBuilder().setCurrentGeoCell(currentLocation))
            .build());

    // notify the websocket
    messaging.broadcastDriverSimulationEvent(
        WebsocketDriverEvent.builder()
            .currentLocation(currentLocation)
            .driverId(driverId)
            .driverStatus(WebsocketDriverEvent.DriverStatus.ENROUTE)
            .ride(rideInformation)
            .build());

    return WebsocketDriverEvent.DriverStatus.ENROUTE;
  }

  @Override
  public String id() {
    return driverId;
  }

  @Override
  public boolean isDone() {
    return stateMachine.isAtTerminalState();
  }

  @Override
  public void apply(Object event) {
    stateMachine.apply(event);
  }

  @Override
  public boolean needReschedule() {
    return true; // the driver never rests.
  }
}

File: statefun-go-sdk/examples/rideshare/driver.go
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"context"
	"fmt"
	"github.com/apache/flink-statefun/statefun-go-sdk/v2/pkg/flink/statefun"
	"github.com/apache/flink-statefun/statefun-go-sdk/v2/pkg/flink/statefun/io"
	"google.golang.org/protobuf/types/known/anypb"
)

type Driver struct {
	Ride     statefun.State `state:"ride"`
	Location statefun.State `state:"location"`
}

func (d *Driver) Invoke(ctx context.Context, out statefun.Output, msg *anypb.Any) error {
	pickup := &PickupPassenger{}
	if err := msg.UnmarshalTo(pickup); err == nil {
		return d.pickupNeeded(ctx, out, pickup)
	}

	driverMessage := InboundDriverMessage{}
	if err := msg.UnmarshalTo(&driverMessage); err == nil {
		if rideStarted := driverMessage.GetRideStarted(); rideStarted != nil {
			return d.whenRideStarted(ctx, out)
		}

		if rideEnded := driverMessage.GetRideEnded(); rideEnded != nil {
			return d.whenRideEnded(out)
		}

		if locationUpdate := driverMessage.GetLocationUpdate(); locationUpdate != nil {
			return d.whenLocationUpdated(out, locationUpdate)
		}
	}

	return nil
}

func (d *Driver) pickupNeeded(ctx context.Context, out statefun.Output, pickup *PickupPassenger) error {
	var ride CurrentRide

	if exists, err := d.Ride.Get(&ride); err != nil {
		return err
	} else if exists {
		// this driver is currently in a ride and therefore cannot take
		// anymore passengers

		driverId := statefun.Self(ctx).Id
		ride := statefun.Caller(ctx)

		return out.Send(*ride, &DriverRejectsPickup{
			DriverId: driverId,
			RideId:   ride.Id,
		})
	}

	// We are called by the ride function, so we remember its id
	// for future communication.

	ride.RideId = statefun.Caller(ctx).Id
	if err := d.Ride.Set(&ride); err != nil {
		return err
	}

	var location CurrentLocation
	if _, err := d.Location.Get(&location); err != nil {
		return err
	}

	// We also need to unregister ourselves from the current
	// geo cell we belong to
	cell := statefun.Address{
		FunctionType: GeoCellType,
		Id:           fmt.Sprint(location.Location),
	}

	if err := out.Send(cell, &LeaveCell{}); err != nil {
		return err
	}

	// Reply to the ride saying we are taking this passenger

	driverId := statefun.Self(ctx).Id
	join := &DriverJoinsRide{
		DriverId:       driverId,
		PassengerId:    ride.RideId,
		DriverLocation: location.Location,
	}

	if err := out.Send(*statefun.Caller(ctx), join); err != nil {
		return err
	}

	record := io.KafkaRecord{
		Topic: "to-driver",
		Key:   driverId,
		Value: &OutboundDriverMessage{
			DriverId: driverId,
			Message: &OutboundDriverMessage_PickupPassenger_{
				PickupPassenger: &OutboundDriverMessage_PickupPassenger{
					RideId:           ride.RideId,
					StartGeoLocation: pickup.PassengerStartCell,
					EndGeoLocation:   pickup.PassengerEndCell,
				},
			},
		},
	}

	if message, err := record.ToMessage(); err != nil {
		return err
	} else {
		return out.SendEgress(ToDriverEgress, message)
	}
}

func (d *Driver) whenRideStarted(ctx context.Context, out statefun.Output) error {
	var location CurrentLocation
	if _, err := d.Location.Get(&location); err != nil {
		return err
	}

	var ride CurrentRide
	if _, err := d.Ride.Get(&ride); err != nil {
		return err
	}

	started := RideStarted{
		DriverId:      statefun.Self(ctx).Id,
		DriverGeoCell: location.Location,
	}

	address := statefun.Address{
		FunctionType: RideType,
		Id:           ride.RideId,
	}

	return out.Send(address, &started)
}

func (d *Driver) whenRideEnded(out statefun.Output) error {
	var ride CurrentRide
	if _, err := d.Ride.Get(&ride); err != nil {
		return nil
	}

	rideAddress := statefun.Address{
		FunctionType: RideType,
		Id:           ride.RideId,
	}

	if err := out.Send(rideAddress, &RideEnded{}); err != nil {
		return err
	}

	d.Ride.Clear()

	// register the driver as free at the current dLocation
	var location CurrentLocation
	if _, err := d.Location.Get(&location); err != nil {
		return err
	}

	geoCell := statefun.Address{
		FunctionType: GeoCellType,
		Id:           fmt.Sprint(location.Location),
	}

	return out.Send(geoCell, &JoinCell{})
}

func (d *Driver) whenLocationUpdated(runtime statefun.Output, update *InboundDriverMessage_LocationUpdate) error {
	var location CurrentLocation
	if exists, err := d.Location.Get(&location); err != nil {
		return err
	} else if !exists {
		// this is the first time this driver gets
		// a location update so we notify the relevant
		// geo cell function.

		location.Location = update.CurrentGeoCell

		geoCell := statefun.Address{
			FunctionType: GeoCellType,
			Id:           fmt.Sprint(location.Location),
		}

		return runtime.Send(geoCell, &location)
	}

	if location.Location == update.CurrentGeoCell {
		return nil
	}

	location.Location = update.CurrentGeoCell
	return d.Location.Set(&location)
}
